<div style="border:solid purple 2px; padding: 20px">

Евгений привет! 👋 Поздравляю с первым выполненным проектом 🙂
    
Меня зовут Рустам Муртазин, и я буду делать ревью проектной работы по предобработке данных. Давай будем общаться на **«ты»**. Если это неприемлемо, обязательно напиши мне в комментариях — мы перейдем на **«вы»**.

Ты проделал большую работу над проектом, однако в любом деле всегда можно можно что-то улучшить. Моя основная цель — не показать, что где-то совершены какие-либо ошибки, а поделиться своим опытом, который поможет тебе в дальнейших работах. Далее в файле ты сможешь увидеть мои комментарии, постарайся учесть их при выполнении следующих заданий.    

Обрати внимание в первую очередь на те, что помечаны <span style="color:red">красным цветом</span>. После их доработки проект будет принят. <span style="color:green">Зеленым цветом</span> отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. <span style="color:orange">Оранжевым цветом</span> выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно). Также в проекте могут быть небольшие «лайфхаки» по Python, не относящиеся к проекту, их я выделил в фиолетовую рамку)

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее».

---

У тебя получилась отличная работа, особенно для первого проекта. Ты показываешь прекрасное владение изучаемым материалом: уверенно пользуешься pandas, умеешь подготавливать, очищать, обогощать данные.  На протяжении всей работы чувствуется глубина проработки задачи и это безусловный плюс для нас.

Как ни старался, критичных проблем я не нашел, ты точно готов идти дальше. Успехов в дальнейшем порхождении курса! 😊

![](https://i.gifer.com/7V3.gif)

<div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Ты уже наверное заметил, недавно добавили плагин `toc` для Jupyter, позволяющий сделать [автоматическое оглавление](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/toc2/README.html). Так вот, если тебе не нравится автоматическая нумерация с `toc`, то ее можно убрать. Для этого нужно сделать следующее:
    
- выведи оглавление, для этого нажми на значок списка (находится справа от `Validate`)
- слева выйдет само оглаление, нажми на шестиренку
- и убери галку с функции автоматической нумерации `Automatically number headings`

In [1]:
#Описание данных

    #children — количество детей в семье
    #days_employed — общий трудовой стаж в днях
    #dob_years — возраст клиента в годах
    #education — уровень образования клиента
    #education_id — идентификатор уровня образования
    #family_status — семейное положение
    #family_status_id — идентификатор семейного положения
    #gender — пол клиента
    #income_type — тип занятости
    #debt — имел ли задолженность по возврату кредитов
    #total_income — ежемесячный доход
    #purpose — цель получения кредита
    
import pandas as pd
import numpy as np
from pymystem3 import Mystem
from nltk.stem import SnowballStemmer 
from collections import Counter
bk = pd.read_csv('/datasets/data.csv')
bk = pd.DataFrame(bk)
bk

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Обрати, пожалуйста, внимание на использование `head`, `tail` или `sample` при форматировании датафреймов. Не забывай про них. Если оставить эту таблицу "как есть", то при сохранении ноутбука, например, на github, таблица развернется во всю длину и займет значительную часть отчета.

In [2]:
#исследуем краткую сводку данных
bk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
#исследуем пропущенные значения по среднему
bk.isna().mean()

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

In [4]:
#переименуем столбцы в более понятный и читабельный вид
bk.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

In [5]:
#days_employed  > 'experience'  общий трудовой стаж в днях
#dob_years > 'age'  возраст клиента в годах
#income_type' > profession '   тип занятости
#total_income > 'income'  ежемесячный доход 


bk.columns = ['children', 'experience', 'age', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'profession', 'debt',
       'income', 'purpose']
bk

children     experience  age education  education_id     family_status  \
0             1   -8437.673028   42    высшее             0   женат / замужем   
1             1   -4024.803754   36   среднее             1   женат / замужем   
2             0   -5623.422610   33   Среднее             1   женат / замужем   
3             3   -4124.747207   32   среднее             1   женат / замужем   
4             0  340266.072047   53   среднее             1  гражданский брак   
...         ...            ...  ...       ...           ...               ...   
21520         1   -4529.316663   43   среднее             1  гражданский брак   
21521         0  343937.404131   67   среднее             1   женат / замужем   
21522         1   -2113.346888   38   среднее             1  гражданский брак   
21523         3   -3112.481705   38   среднее             1   женат / замужем   
21524         2   -1984.507589   40   среднее             1   женат / замужем   

       family_status_id gender profession  debt         income  \
0                     0      F  сотрудник     0  253875.639453   
1                     0      F  сотрудник     0  112080.014102   
2                     0      M  сотрудник     0  145885.952297   
3                     0      M  сотрудник     0  267628.550329   
4                     1      F  пенсионер     0  158616.077870   
...                 ...    ...        ...   ...            ...   
21520                 1      F  компаньон     0  224791.862382   
21521                 0      F  пенсионер     0  155999.806512   
21522                 1      M  сотрудник     1   89672.561153   
21523                 0      M  сотрудник     1  244093.050500   
21524                 0      F  сотрудник     0   82047.418899   

                            purpose  
0                     покупка жилья  
1           приобретение автомобиля  
2                     покупка жилья  
3        дополнительное образование  
4                   сыграть свадьбу  
...                             ...  
21520             операции с жильем  
21521          сделка с автомобилем  
21522                  недвижимость  
21523  на покупку своего автомобиля  
21524         на покупку автомобиля  

[21525 rows x 12 columns]

In [6]:
#посмотрим общую информацию по числовым значениям
bk.describe()

children     experience           age  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt        income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [7]:
#и нечисловым признакам
bk.describe(include=['object', 'bool'])

education    family_status gender profession  purpose
count      21525            21525  21525      21525    21525
unique        15                5      3          8       38
top      среднее  женат / замужем      F  сотрудник  свадьба
freq       13750            12380  14236      11119      797

In [8]:
#судя по таблице нечисловых признаков, в столбце уровня образования возможно есть ошибки,
#по гендерному признаку появился третий человеческий пол, а жениться стали так часто что самой
#распростарненной целью кредита стала "свадьба". Чтобы в дальнейшем навести порядок в данных
#таблицы, сразу посмотрим эти загадочные значения на предмет ошибок

bk['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Старайся промежуточные выводы/комментарии оформлять также в ячейках `Markdown`, это повышает читаемость кода. 

</div>

In [9]:
#проверяем уникальность пола клиентов
bk['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [10]:
#проверяем уникальность цели кредитов
bk['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

**===========================================================================================================**
**Вывод**

Первые наблюдения:
1. В информации о количестве детей: непонятное минимальное значение "-1  и максимальное "20", скорее всего это ошибки; 
2. Данные о стаже содержат отрицательные значения; есть пропущенные строки; имеется плавающая точка - стоит присвоить целочисленные значения; среднее значение - 63046 дней (или 173 года) - видимо у кого-то не верно заведенная информация; подсчет идет в днях - лучше перевести в года;
3. Минимальное нулевое значение в данных о возрасте, говорит о том, что не у всех клиентов указана эта информация, нужно исследовать и исправить это недоразумение;
4. Данные в столбце уровня образования имеют разные регистры, это объясняет множество уникальных значений. Судя по максимальному количеству id их должно быть 5.
5. Надо разобраться с одним клиентом, который не смог определиться со свои полом. 
6. В информации о доходах есть пропущенные строки, непонятна единица измерения в виде плавающей точки - стоит присвоить целочисленные значения;
7. Цель кредита имеет похожие друг на друга записи, нужно провести лемматизацию.

Над чем будем работать:
1. Обработка пропусков (и странных значений) - исследуем и исправим пропущенные значения в данных о стаже и доходе; проверим отрицательные значения в стаже; приведем регистр столбца уровня образования к единому формату; исправим непонятные - и + в данных о детях; обработаем нулевой возраст; 
2. Замена типа данных - данным о стаже и доходе необходимо привести целочисленные значения
3. Обработка дубликатов - найти, исследовать и обработать при наличии
4. Лемматизация - цели кредитов найдем и присвоим уникальные значения
5. Категоризация данных - чтобы ответить на поставленные вопросы как минимум группировку необходимо будет сделать по количесту детей, семейному положению, уровню дохода и цели кредита. 
**===========================================================================================================**

## Шаг 2. Предобработка данных

### Обработка пропусков

In [11]:
#1. исследуем и исправим пропущенные значения в данных о стаже и доходе. Скорее всего, отсутствующие
#данные о стаже говорят о том, что клиент возможно никогда еще не работал, а отсутствие дохода это
#только подтверждает, проверим совпадают ли пропущенные строки

bk[(bk['income'].isnull() == True) & (bk['experience'].isnull() == True)].info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
children            2174 non-null int64
experience          0 non-null float64
age                 2174 non-null int64
education           2174 non-null object
education_id        2174 non-null int64
family_status       2174 non-null object
family_status_id    2174 non-null int64
gender              2174 non-null object
profession          2174 non-null object
debt                2174 non-null int64
income              0 non-null float64
purpose             2174 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


In [12]:
#гипотеза подтвердилась, у клиентов с отсутсвующем стажем дохода также нет.
#для общего понимания посмотрим кто эти клиенты по профессии

bk[(bk['income'].isnull() == True) & (bk['experience'].isnull() == True)]['profession'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: profession, dtype: int64

In [13]:
#в большинстве категорий есть такие клиенты, скорее всего данные были просто не указаны.
#предположу, что пропущенные строки в виде стажа и дохода просто не были официально подтверждены,
#поэтому логично, что в базе их нет.

#поясню - 
#специфика формирования подобной базы заключается в том, что данные берутся из скорринговой анкеты,
#которую заполняет непосредственно сам клиент, либо, с его слов кредитный специалист. Но, как
#известно, банки, в более выгодных для клиента кредитных продуктах, запрашивают подтверждение
#данных: трудовой стаж подтверждается заверенной копией трудовой книжки,  доход – справкой от
#работодателя по форме 2 ндфл. И зачастую клиент приносит сразу оба документа. Такое подтверждение
#увеличивает шансы клиента на получение кредита, а банк видит его реальную платежеспособность. 
#Поэтому пропущенные значение необходимо заменить на нули, а не подставлять среднее или медианное
#значение, так как это повысит шанс получения кредита, а не качество данных.

#в таком случае заменим пропуски на нули.
bk['experience'] = bk['experience'].fillna(0)

In [14]:
bk['income'] = bk['income'].fillna(0)

<div class="alert alert-warning">
<h2> Комментарий ревюьера <a class="tocSkip"> </h2> 

Может стоить попробовать заменить пропуски более детально, например заменяя на медианное значение по группам (для `income` например группируя по `profession`). Данный навык точно пригодится в последующих проектах 😉 
    
P.S Комментарий не актуален, вижу, как ты обработал нули в `age`

</div>

In [15]:
#проверим замену
bk.isnull().sum()

children            0
experience          0
age                 0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
profession          0
debt                0
income              0
purpose             0
dtype: int64

In [16]:
#все пропуски были изменены на нули
#теперь разберемся с отрицательными значениями в стаже. Скорее всего это техническая ошибка при
#выгрузке базы, другого объяснения не вижу, но исправлять надо. Исправим на абсолютные значения.
bk['experience'] = bk['experience'].abs()
bk['experience'].describe()

count     21525.000000
mean      60156.419005
std      133355.929525
min           0.000000
25%         610.652074
50%        1808.053434
75%        4779.587738
max      401755.400475
Name: experience, dtype: float64

In [17]:
#теперь смущает максимальное значение на фоне среднего. Для визуального удобства переведем дни
#стажа в годы и поместим их в новый столбец

bk['years_experience'] = bk['experience'] / 365

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip">  </h2>

Мне нравится ход твоих мыслей и детальная проработка 👍

</div>

In [18]:
bk['years_experience'] = bk['years_experience'].astype(int)

In [19]:
#проверим
bk['years_experience'].head(5)

0     23
1     11
2     15
3     11
4    932
Name: years_experience, dtype: int64

In [20]:
#судя по выводу в 932 года в стаже клиентов также закрались ошибки. Посмотрим сколько клиентов у нас
#вышли за рамки возможного, предполагая что клиент начал работать с 16 лет и вышел на пенсию в 60.

mistakes = bk[bk['years_experience'] > 44].count()[0]
print("Количество клиентов со стажем более 44 лет: ", mistakes)
share_mistakes = mistakes / bk['years_experience'].count()
print("Доля ошибок в списке: " '{:.2%}'.format(share_mistakes))

Количество клиентов со стажем более 44 лет:  3448
Доля ошибок в списке: 16.02%


<div style="border:solid purple 5px; padding: 20px">
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>
    
<h3> Библиотека os <a class="tocSkip"> </h3>

Модуль os в Python — это библиотека функций для работы с операционной системой. Методы, включенные в неё позволяют определять тип операционной системы, получать доступ к переменным окружения, управлять директориями и файлами:

- проверка существования объекта по заданному пути;
- определение размера в байтах;
- удаление;
- переименование и др.

Библиотека довольно обширная, при желании, можно почитать [официальную документацию](https://docs.python.org/3/library/os.html), либо короткий [перевод](https://pythonworld.ru/moduli/modul-os.html). Вот некоторые методы этой библиотеки

Чтобы узнать, какая она на вашей операционной системе, используется функция `getcwd`

![](https://i.ibb.co/4RYXGqV/image.png)

Функция `chdir` сменит ее на другую, которая будет указана аргументом

![](https://i.ibb.co/rwM0WY8/image.png)

Функция `listdir` без указания аргумента покажет все файлы и папки текущей рабочей директории. Можно также указать аргументом интересующий вас путь

![](https://i.ibb.co/4pWQ16Y/image.png)

Чтобы создать папку, применяется функция `mkdir`. Эта функция сможет создать только одну папку в существующем пути, если указанного пути не существует, то вызовется ошибка. Если папка уже существует - вызовется ошибка

![](https://i.ibb.co/q53Z6YD/image.png)

In [21]:
#достаточно много для технической ошибки, пока непонятно что с этим делать, вернемся к этим данным
#после обработки дубликатов и лемматизации

#теперь приведем значения в уровне образования к единому регистру
bk['education'] = bk['education'].str.lower()
bk['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [22]:
#как и должно быть, получилось пять уникальных уровней.

#далее будем разираться с количеством детей. Посмотрим сначала на количестсво непонятных значений.
print("Количество с  0: ", bk[bk['children'] == 0].count()[0])
print("Количество с  -1: ", bk[bk['children'] == -1].count()[0])
print("Количество с  20: ", bk[bk['children'] == 20].count()[0])

Количество с  0:  14149
Количество с  -1:  47
Количество с  20:  76


In [23]:
#снова наблюдаются ошибки, данные по 47 клиентам о количестве детей заменим на 0,
#анализ от этого не пострадает, а по 76 клиентам изменим знчения на 2 - согласно статистики,
# из открытызх источников инфорации в интренете, в РФ нет столько семей с 20 детьми. 

bk['children'] = bk['children'].replace(-1, 1)
bk['children'] = bk['children'].replace(20, 2)
bk['children'].unique()

array([1, 0, 3, 2, 4, 5])

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Очень правильное замечание и отлично применен метод `replace` для замены

</div>

In [24]:
#далее на очереди возраст клиентов, необходимо исследовать нулевые данные

#распределим клиентов по возрасту устоявшимся категориям - до 18 лет, от 18 до 65,
print("Количество клиентов в возврсте 0: ", bk[bk['age'] == 0].count()[0])
print("Количество клиентов в возврсте до 18: ", bk[bk['age'] < 18].count()[0])
print("Количество клиентов в возврсте от 18: ", bk[bk['age'] >= 19].count()[0])

Количество клиентов в возврсте 0:  101
Количество клиентов в возврсте до 18:  101
Количество клиентов в возврсте от 18:  21424


In [25]:
#как видно из расчетов, "нулевые" клиенты и только достигшие совершеннолетие - это один и теже люди.
#переведем эти данные в тип "int" и сразу проверим результат на количество нулей
bk['age'] = bk.groupby('profession')['age'].transform(lambda x: x.replace(0, int(x.mean())))
print('Результат: ', bk[bk['age'] == 0].count()[0])

Результат:  0


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Отличная замена, тогда оранжевый комментарий отпадает, вижу, что ты владеешь методом

**===========================================================================================================**
**Вывод:** 
1.	После обработки пропущенных значений в стаже и доходе клиента, выяснилось, что данные пропуски имеют одни и те же люди. Отсутствующая информация объясняется, скорее всего, тем, что в базу попали клиенты, не подтвердившие официально стаж работы и свои доходы, а в базу можно заносить только официально подтвержденные данные.  Пропуски заменил на 0. 
2.	Отрицательные значения в стаже возникли, возможно, в результате технической ошибки при выгрузке базы, либо, при занесении непосредственно самих данных. Все отрицательные значения привел к абсолютным. 
3.	Стаж работы из  формата в «в днях» для более удобного визуального понимания картины, перевел «в года». Однако после перевода обнаружились клиенты, которые имеют стаж работы более максимально официально возможного 44 лет. И таких суммарно набралось ≈16% от общего числа.  Это слишком много для ошибки в выгрузке. Возможно ситуация прояснится после обработки дубликатов.
4.	Данные в столбце уровня образования привел к единому регистру, теперь количество уникальных значений соответствует количеству уникальных id. 
5.	Данные возраста клиентов также подверглись изменениям в результате выявленных ошибок: 0,5% от общего числа строк не сильно повлияют на общий результат, поэтому отрицательные значения было принято решение перевести в  абсолютные, а  клиентов с количеством детей «20» в «2».
6.	В возрасте клиентов «нулевых» оказалось ровно столько же сколько и несовершеннолетних из всего списка. Либо, это совпадение и все данные клиенты подавали заявки на кредит до 18 лет, либо, это очередные ошибки при занесении информации. Все 0 были приведены в целые числа. 
**===========================================================================================================**

### Замена типа данных

In [26]:
#еще раз посмотрим на имеющиеся в столбцах типы данных
bk.dtypes

children              int64
experience          float64
age                   int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
profession           object
debt                  int64
income              float64
purpose              object
years_experience      int64
dtype: object

In [27]:
#переведем данные в удобный формат 
bk['experience'] = bk['experience'].astype('int') 
bk['income'] = bk['income'].astype('int')
bk.dtypes

children             int64
experience           int64
age                  int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
profession          object
debt                 int64
income               int64
purpose             object
years_experience     int64
dtype: object

**===========================================================================================================**
**Вывод**

Все данные представлены в довольно удобном формате. Для замены типа данных был использован метод 'astype', так как мы уже обработали пропущенные значения (иначе можно было бы использовать 'to_numeric'). Месячную зарплату и стаж в днях перевели в 'int', чтобы видеть целочисленные значения. 
**===========================================================================================================**

### Обработка дубликатов

In [28]:
#посчитаем количество дубликатов в массиве.
print('Число дубликатов в массиве данных:', bk.duplicated().sum())
print('Число полностью идентичных строк:', bk.duplicated(keep = False).sum())
print('Доля дубликатов из общей длины массива: {:.2%}'.format(bk.duplicated().sum() / len(bk)))

Число дубликатов в массиве данных: 71
Число полностью идентичных строк: 137
Доля дубликатов из общей длины массива: 0.33%


In [29]:
#для наглядности выведем первые 5 строк полностью "идентичных" клиентов.
bk[bk.duplicated(keep = False)].sort_values('age', ascending = False).head()

children  experience  age education  education_id     family_status  \
7938         0           0   71   среднее             1  гражданский брак   
9604         0           0   71   среднее             1  гражданский брак   
6537         0           0   71   среднее             1  гражданский брак   
5865         0           0   66   среднее             1    вдовец / вдова   
9528         0           0   66   среднее             1    вдовец / вдова   

      family_status_id gender profession  debt  income  \
7938                 1      F  пенсионер     0       0   
9604                 1      F  пенсионер     0       0   
6537                 1      F  пенсионер     0       0   
5865                 2      F  пенсионер     0       0   
9528                 2      F  пенсионер     0       0   

                              purpose  years_experience  
7938            на проведение свадьбы                 0  
9604            на проведение свадьбы                 0  
6537            на проведение свадьбы                 0  
5865  операции со своей недвижимостью                 0  
9528  операции со своей недвижимостью                 0

In [30]:
#удалим одинаковые строки.
bk = bk.drop_duplicates().reset_index(drop = True)

#проверим число дубликатов.
print('Число дубликатов в массиве данных:', bk.duplicated().sum())

Число дубликатов в массиве данных: 0


**===========================================================================================================**
**Вывод**

При обработке данных были найдены дубликаты в количестве 71 шт., что составляет 0,33% от всего числа клиентов в массиве данных. Необходимо выяснить, это задвоение данных/ошибочное создание дублей, или же ошибка носит человеческий фактор. При удалении дубликатов использовался метод drop_duplicates(), так как его целью является поиск идентичных строк, что нам и требовалось.
**===========================================================================================================**

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

![](https://i.gifer.com/R0YH.gif)

### Лемматизация

In [31]:
#в таблице есть столбец, с данными о цели кредита, но как мы уже ранее выяснили, они имеют похожие
#друг на друга значения. Исследуем количество уникалных лемм и определим категории целей. 

#подсчитаем все варианты целей кредита.
bk['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Повторюсь, выноси все логические комментарии в Markdown, обычно такие работы я отправляю на доработку, переносить комментарии, но у тебя по остальным пунктам сделано все идеально и не хочется только из-за этого не принимать работу

In [32]:
#cреди всех возможных целей получения кредита выделим ключевые слова, которые отражают главную цель.

m = Mystem()
lemmas = []
for element in bk['purpose']:
    lemma = m.lemmatize(element)
    lemmas.extend(lemma)

unique_lemmas = Counter(lemmas)
sorted(unique_lemmas.items(), key = lambda pair: pair[1], reverse=True)

[(' ', 33570),
 ('\n', 21454),
 ('недвижимость', 6351),
 ('покупка', 5897),
 ('жилье', 4460),
 ('автомобиль', 4306),
 ('образование', 4013),
 ('с', 2918),
 ('операция', 2604),
 ('свадьба', 2324),
 ('свой', 2230),
 ('на', 2222),
 ('строительство', 1878),
 ('высокий', 1374),
 ('получение', 1314),
 ('коммерческий', 1311),
 ('для', 1289),
 ('жилой', 1230),
 ('сделка', 941),
 ('дополнительный', 906),
 ('заниматься', 904),
 ('проведение', 768),
 ('сыграть', 765),
 ('сдача', 651),
 ('семья', 638),
 ('собственный', 635),
 ('со', 627),
 ('ремонт', 607),
 ('подержанный', 486),
 ('подержать', 478),
 ('приобретение', 461),
 ('профильный', 436)]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Реализация правильная, при этом предлагаю более короткое и быстрое решение с использованием [join()](https://stackoverflow.com/questions/5618878/how-to-convert-list-to-string)
    
    lemmas = m.lemmatize(' '.join(bk['purpose']))
    print(Counter(lemmas))

</div>

In [33]:
#выделим категории из результата, основные ключевые слова можно выделить вручную

categories = ["сдача", "коммерческий", "жилье", "образование", "свадьба", "недвижимость", "автомобиль"]
categories

['сдача',
 'коммерческий',
 'жилье',
 'образование',
 'свадьба',
 'недвижимость',
 'автомобиль']

In [34]:
# проведем лемматизацию,
# одновременно заменив полученный список лемм в каждой строке на главное ключевое слово из списка категорий

def lemmatize(text):
    lemma = m.lemmatize(text)
    for word in categories:
        if word in lemma:
            lemma = word
    return lemma

bk['purpose_group'] = bk['purpose'].apply(lemmatize)        
bk.head(5)

children  experience  age education  education_id     family_status  \
0         1        8437   42    высшее             0   женат / замужем   
1         1        4024   36   среднее             1   женат / замужем   
2         0        5623   33   среднее             1   женат / замужем   
3         3        4124   32   среднее             1   женат / замужем   
4         0      340266   53   среднее             1  гражданский брак   

   family_status_id gender profession  debt  income  \
0                 0      F  сотрудник     0  253875   
1                 0      F  сотрудник     0  112080   
2                 0      M  сотрудник     0  145885   
3                 0      M  сотрудник     0  267628   
4                 1      F  пенсионер     0  158616   

                      purpose  years_experience purpose_group  
0               покупка жилья                23         жилье  
1     приобретение автомобиля                11    автомобиль  
2               покупка жилья                15         жилье  
3  дополнительное образование                11   образование  
4             сыграть свадьбу               932       свадьба

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Советую здесь убрать строку `lemma = word` и сразу возвращать lemma, т.е так
    
    def lemmatize(text):
        lemma = m.lemmatize(text)
        for word in categories:
            if word in lemma:
                return lemma
        
В таком случае, если первое условие True, функция прерывается и остальные условия не проверяются. Это сильно сыграет на быстродействии функции

In [35]:
# Посчитаем все возможные категории целей кредита.
bk['purpose_group'].value_counts()

недвижимость    5040
автомобиль      4306
образование     4013
жилье           3809
свадьба         2324
коммерческий    1311
сдача            651
Name: purpose_group, dtype: int64

In [36]:
#поскольку жилье тоже относится к недвижимости, то сделаем соответствующую замену.
#кроме того, заменим категории по покупке коммерческой недвижимости или покупке жилой недвижимости 
#для сдачи в наем на инвестиционную цель и снова посчитаем число категорий.
bk.loc[bk['purpose_group'] == 'жилье', 'purpose_group'] = 'недвижимость'
bk.loc[(bk['purpose_group'] == 'коммерческий') | (bk['purpose_group'] == 'сдача'), 'purpose_group'] = 'инвестиционная цель'
bk['purpose_group'].value_counts()

недвижимость           8849
автомобиль             4306
образование            4013
свадьба                2324
инвестиционная цель    1962
Name: purpose_group, dtype: int64

**===========================================================================================================**
**Вывод**


Из цели получения кредита каждого клиента было выделено по одному ключевому слову:

    - цель на покупку своего автомобиля заменяется на автомобиль
    - цель покупка жилой недвижимости заменяется на недвижимость
    и т.д.

Подобное выделение ключевых слов из целей клиентов упрощает дальнейший анализ и сохраняет основной смысл цели получения кредита.

Все цели, содержащие производные слова от слова жилье (3809 чел.), были заменены на категорию недвижимость (цели с исходным ключевым словом насчитывают 5040 чел.), кроме целей по покупке коммерческой недвижимости (1311 чел.) или покупке жилой недвижимости для сдачи в наем (651 чел.).

Если обычная покупка жилой недвижимости относится к конечному потреблению, то приобретение коммерческой недвижимости или недвижимости для сдачи в наем относится к инвестиционным целям, т.е. этот кредит клиенты планируют не только вернуть, но вероятно еще и окупить. То есть клиенты планируют в ближайшем будущем получать положительный денежный поток. Доля таких клиентов составляет 9,1% от общего числа клиентов, поэтому ими не стоит пренебрегать, а стоит выделить отдельно.
**===========================================================================================================**

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Жирный плюс! Это один из сложнейших этапов проекта, справился на отлично

</div>

### Категоризация данных

**- Группировка по наличию детей**

In [37]:
#ранее после лематизации уже была выполнена группировка клиентов по цели кредита, и добавлен в таблицу столбец 
#с новыми данными по категориям. Далее сгруппируем клиентов по количеству детей, семейному положению и доходу

#посчитаем клиентов по количеству детей
children_group = pd.DataFrame()
children_group['count_children'] = bk.groupby('children')['debt'].count()
children_group['share_children'] = children_group['count_children'] * 100 / bk['children'].count()
children_group['share_children'] = round(children_group['share_children'], 2).astype(str) + '%'
children_group

count_children share_children
children                               
0                  14091         65.68%
1                   4855         22.63%
2                   2128          9.92%
3                    330          1.54%
4                     41          0.19%
5                      9          0.04%

In [38]:
#так как по заданию необходимо узнать как влияет наличие детей на возврат кредита в срок, 
#выделим две соответствующие группы и добавим их в таблицу новым столбцом
def children_count(children):
    if children == 0:
            return 'Без детей'
    return 'С детьми'

In [39]:
bk['children_count'] = bk['children'].apply(children_count)


In [40]:
#сделаем сводную повторно
children_count_group = pd.DataFrame()
children_count_group['count_children'] = bk.groupby('children_count')['debt'].count()
children_count_group['share_children'] = children_count_group['count_children'] * 100 / bk['children_count'].count()
children_count_group['share_children'] = round(children_count_group['share_children'], 2).astype(str) + '%'
children_count_group

count_children share_children
children_count                               
Без детей                14091         65.68%
С детьми                  7363         34.32%

**- Группировка по семейному положению**

In [41]:
#посчитаем клиентов по семейному положению

family_group = pd.DataFrame()
family_group['count_family'] = bk.groupby('family_status')['debt'].count()
family_group['share_family'] = family_group['count_family'] * 100 / bk['family_status'].count()
family_group['share_family'] = round(family_group['share_family'], 2).astype(str) + '%'
family_group = family_group.sort_values(by ='count_family', ascending = 0)
family_group

count_family share_family
family_status                                   
женат / замужем               12339       57.51%
гражданский брак               4151       19.35%
Не женат / не замужем          2810        13.1%
в разводе                      1195        5.57%
вдовец / вдова                  959        4.47%

**- Группировка по доходу**

In [42]:
#в качестве границ для категоризации доходов используем квантили (25%, 50% и 75%).

bk['income'].quantile([0.25,0.5,0.75])

0.25     89088.50
0.50    135781.00
0.75    195813.25
Name: income, dtype: float64

In [43]:
#с помощью функции определим категории и добавим их новым столбцом в таблицу

def income_status(income):
    if income <= 89088.50:
            return 'Низкий уровень'
    if income <= 135781.0:
            return 'Средний уровень'
    if income < 195813.25:
            return 'Высокий уровень'
    return 'Сверхвысокий уровень'

In [44]:
bk['income_status'] = bk['income'].apply(income_status)

In [45]:
#посчитаем клиентов по уровню дохода
income_status_group = pd.DataFrame()
income_status_group['count_status'] = bk.groupby('income_status')['debt'].count()
income_status_group['share_income'] = income_status_group['count_status'] * 100 / bk['income_status'].count()
income_status_group['share_income'] = round(income_status_group['share_income'], 3).astype(str) + '%'
income_status_group

count_status share_income
income_status                                  
Высокий уровень               5363      24.998%
Низкий уровень                5364      25.002%
Сверхвысокий уровень          5364      25.002%
Средний уровень               5363      24.998%

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Здесь, кстати для деления по доходу можно было применить метод [cut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html) или [qcut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html)(в этом случае деление будет проводиться по проценилям). Обрати внимание, у данного метода есть параметр `labels`, так можно оформить группы красивее 😊

</div>

**===================================================================================================**
**Вывод**

Сгруппировав данные по столбцам, можно сделать вывод, что в данном банке чаще всего заявки на кредит подают молодые семьи с не более одним ребенком на улучшение жилищниых условий, скорее всего это ипотеки с первоначальным взносом от материнского капитала. В категориях по уровню дохода клиентов практически одинаково. 
**===================================================================================================**

## Шаг 3. Ответьте на вопросы

**- Есть ли зависимость между наличием детей и возвратом кредита в срок?**

In [46]:
#определим какое количестсво клиентов с детьми и без возвращают кредит в срок
children_pivot = bk.pivot_table(index = ['children_count'], \
                                columns = ['debt'], \
                                values = 'purpose', aggfunc='count')

children_pivot['ratio'] = children_pivot[1] / children_pivot[0] * 100
children_pivot

debt                0     1      ratio
children_count                        
Без детей       13028  1063   8.159349
С детьми         6685   678  10.142109

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

У тебя отличная работа!
    
Может тебе будет интересно: для улучшения визуализации датафрейма, можно пользоваться [style](https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html) библиотеки `pandas`. Там много прикольных штук для того, чтобы сделать датафрейм красивее 😊
Например, в нашем случае, чтобы выводить результат в процентах, а не в долях, можно прописать следующее:
    
`children_pivot.style.format({'ratio': '{:.2%}'})`  
    
</div>

In [47]:
#посмотрим как влияет появление детей в семье на платежеспособность
children_pivot_two = bk.pivot_table(index = ['children'], \
                                columns = ['debt'], \
                                values = 'purpose', aggfunc='count')

children_pivot_two['ratio'] = children_pivot_two[1] / children_pivot_two[0] * 100
children_pivot_two = children_pivot_two.sort_values(by ='ratio', ascending = 1)
children_pivot_two

debt            0       1      ratio
children                            
0         13028.0  1063.0   8.159349
3           303.0    27.0   8.910891
1          4410.0   445.0  10.090703
2          1926.0   202.0  10.488058
4            37.0     4.0  10.810811
5             9.0     NaN        NaN

**===================================================================================================**
**Вывод**

Проанализировав полученные данные, можно сделать вывод, что клиенты не имеющие детей ответственнее относятся к выполнению обязательств и выплачивают кредит в срок чаще чем нарушают договор. С появлением в семье детей этот показатель снижается. Скорее всего на это влияет сокращение семейного бюджета, а при сложной материальной ситуации клиент скорее предпочтет задолжать банку и отсрочить платеж, чем не обеспечить своего ребенка всем необходимым. Так как клиенты с детьми оказались более рискованными, при расчете скоррингового проходного бала и процентной ставки банку стоит учитывать данный фактор.
**===================================================================================================**

**- Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [48]:
family_pivot = bk.pivot_table(index=['family_status'], columns=['debt'], values='purpose', aggfunc='count')
family_pivot['ratio'] = family_pivot[1] / family_pivot[0] * 100
family_pivot = family_pivot.sort_values(by ='ratio', ascending = 1)
family_pivot

debt                       0    1      ratio
family_status                               
вдовец / вдова           896   63   7.031250
в разводе               1110   85   7.657658
женат / замужем        11408  931   8.160940
гражданский брак        3763  388  10.310922
Не женат / не замужем   2536  274  10.804416

**===================================================================================================**
**Вывод**

Проанализировав полученные данные, можно сделать вывод, что люди оказавшиеся в одиночестве без своей второй половинки, выплачивают чаще кредиты в срок. Видимо, жизнь в браке учит отдавать долги больше, чем жизнь вне брака, так как в большей степени просрочку допускают клиенты, не обремененные семейными узами.

**===================================================================================================**

**- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [49]:
income_pivot = bk.pivot_table(index=['income_status'], columns=['debt'], values='purpose', aggfunc='count')
income_pivot['ratio'] = income_pivot[1] / income_pivot[0] * 100
income_pivot = income_pivot.sort_values(by ='ratio', ascending = 1)
income_pivot

debt                     0    1     ratio
income_status                            
Сверхвысокий уровень  4981  383  7.689219
Низкий уровень        4945  419  8.473205
Средний уровень       4905  458  9.337411
Высокий уровень       4882  481  9.852519

**===================================================================================================**
**Вывод**

Логично, что люди, имеющие более высокий доход, вовремя исполняют кредитные обязательства. Кроме того, имея лишние сбережения, или, получив внеплановую прибыль от совершения различных сделок, они закрывают кредит досрочно. Но, по полученным данным, на первый взгляд, странным является факт низкого показателя погашения у клиентов с низким уровнем дохода. Объяснить это можно тем, что люди с наименьшими доходами лучше распоряжаются своими средствами и умеют контролировать свой бюджет, чем те, кто зарабатывает немногим больше. 
**===================================================================================================**

<div style="border:solid purple 5px; padding: 20px">
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>
    
<h3> PEP 8 - руководство по написанию кода на Python  <a class="tocSkip"> </h3>

В python есть своя философия. Именно отсюда берутся определенные правила и стандарты. В пайтон такое соглашение назвали PEP 8 (читается - пэп). PEP 8 создан на основе рекомендаций Гуидо ван Россума (создатель языка Python) с добавлениями от Барри Уорсо (соратник).

Единый стиль оформления делает код понятным для самого программиста и его коллег с разным уровнем подготовки. В идеале наиболее сложный фрагмент кода должен быть понятен с первого прочтения. Это упрощает командную разработку и обучение новичков, позволяет вам быстро возвращаться к собственным давним проектам.

Каждый уважающий себя питонист обязан знать этот стандарт. Читайте [оригинал](https://www.python.org/dev/peps/pep-0008/) или [перевод](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html), главное  —  вдумчиво!

Когда проект небольшой, несложно следить за качеством кода, но вот когда проект принимает маштабы ... часто это становиться проблемой. В этом случае выручат скрипты автоматической проверки кода. Кстати, PyCharm проверяет написанное «на лету». А если вы привыкли работать в Jupyter, то на GitHub есть целый раздел [Python Code Quality Authority](https://github.com/PyCQA/), где хранятся утилиты для повышения качества кода, в том числе для проверки стиля на соответствие PEP 8: flake8, pycodestyle, pep8-naming.

![](https://i.ibb.co/Gv0PvHQ/mini-magick20171220-84-vtyy26.png)

**- Как разные цели кредита влияют на его возврат в срок?**

In [50]:
purpose_pivot = bk.pivot_table(index=['purpose_group'], columns=['debt'], values='purpose', aggfunc='count')
purpose_pivot['ratio'] = purpose_pivot[1] / purpose_pivot[0] * 100
purpose_pivot = purpose_pivot.sort_values(by ='ratio', ascending = 1)
purpose_pivot

debt                    0    1      ratio
purpose_group                            
недвижимость         8218  631   7.678267
инвестиционная цель  1811  151   8.337935
свадьба              2138  186   8.699719
образование          3643  370  10.156464
автомобиль           3903  403  10.325391

**===================================================================================================**
**Вывод**

Самая низкая вероятность задолженности у клиентов, которые берут кредит для собственной недвижимости (покупка, строительство и т.д.). Люди настолько заинтересованы в собственном жилье, что платят даже ответственнее по сравнению с теми, кто берет кредит на инвестиционные цели (7,6% против 8.3%). Интересный факт: клиенты, берущие кредит на свадьбу, не самые безответственные плательщики из выборки – для них вероятность всего 8,6%. Хуже них платят только по образовательным и автокредитам. Возможно, такое происходит потому что такие кредиты являются краткосрочными и проценты по ним выше.
**===================================================================================================**

## Шаг 4. Общий вывод

**Консолидировав все выводы по вопросам, можно сделать заключение, что факторы выплаты кредита в срок зависят от:**
- количества иждевенцев в семье: с увеличением количества детей мы видим увеличение количества просроченных задолженностей, хотя люди с 3 детьми чаще платят в срок чем люди с 1 ребенком. Данные не однозначные, возможно, нужна большая выборка, чем та, которую мы имеем. Бездетные, как правило реже просрачивают оплату по кредиту, чем люди с детьми.
- семейного положения: самые низкие показатели у разведенных и вдовцов — можно предположить, что возвращаемость кредитов зависит в том числе от наличия кормильца или супруга в семье.
- уровнем дохода: очевидно, что клиент зарабатывающий больше (чем 75% других клиентов из выборки)  с наименьшей вероятностью окажется в должниках (вероятность всего 7,6%).
- цели кредита: заемщики, берущие кредит для приобретения жилья, наиболее ответственны и менее склонны нарушать обязательства по выплатам кредита в срок.



**Портрет идеального клиента** для банка можно было бы описать следующим образом:
- состоит в браке (либо ранее был женат/замужем), без детей, зарабатывает больше 196 тыс. рублей в месяц, цель получения кредита – приобретение собственного жилья.
Такое описание говорит с одной стороны об умении нести ответственность в некоторой степени (в виде официального брака в настоящем или в прошлом), о высоком уровне достатка и некоторой доле эгоизма, поскольку человек не имеет детей. Таким образом, у него есть средства на выплату кредита, есть собственная материальная цель, но нет сдерживающих факторов в виде иждивенцев.


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2> 

В целом отличный вывод, но здесь еще стоит добавить результаты работ на втором и третьем шагах, как были обрабатаны пропуски, их возможные причины и тд. Также полезно выделить блок «Рекомендации» (с описанием проблем в данных, пропуски, разные регистры и т.д).
    
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [х]  файл изучен;
- [х]  определены пропущенные значения;
- [х]  заполнены пропущенные значения;
- [х]  есть пояснение, какие пропущенные значения обнаружены;
- [х]  описаны возможные причины появления пропусков в данных;
- [х]  объяснено, по какому принципу заполнены пропуски;
- [х]  заменен вещественный тип данных на целочисленный;
- [х]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [х]  удалены дубликаты;
- [х]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [х]  описаны возможные причины появления дубликатов в данных;
- [х]  выделены леммы в значениях столбца с целями получения кредита;
- [х]  описан процесс лемматизации;
- [х]  данные категоризированы;
- [х]  есть объяснение принципа категоризации данных;
- [х]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [х]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [х]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [х]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [х]  в каждом этапе есть выводы;
- [х]  есть общий вывод.